src/typechecker.rs

```diff
@@ -2415,8 +2419,12 @@ pub fn typecheck_expression(
                 .find_struct_in_scope(0, "Set")
                 .expect("internal error: Set builtin definition not found");
 
-            let type_id =
-                project.find_or_add_type_id(Type::GenericInstance(set_struct_id, vec![inner_ty]));
+            // FIXME: Make sure type_hint is of type Set<T>
+            let type_id = if inner_ty == UNKNOWN_TYPE_ID && type_hint.is_some() {
+                type_hint.unwrap()
```

Comment:
This is a little ugly, but I can't think of a nice way to not use `.unwrap()` here without breaking up into something like:
```rust
                let type_id = if let Some(type_hint) = type_hint {
                if inner_ty == UNKNOWN_TYPE_ID {
                    type_hint
                } else {
                    project.find_or_add_type_id(Type::GenericInstance(set_struct_id, vec![inner_ty]))
                }
            } else {
                project.find_or_add_type_id(Type::GenericInstance(set_struct_id, vec![inner_ty]))
            };
```
which doesn't necessarily look that much better